### ISSUE TITLE
How to display contacts of atoms within the same residue?

### ISSUE THREAD

**Issue by sam-mahdi** [dwhswenson/contact_map#129]
I'm comparing contact maps between a single protein in 2 different conformations. I'm looking only at contacts of Nitrogen backbone with the rest of the protein (i.e. query == N, haystack is default). This works with no issues, except I've noticed the contacts are only between atoms of other amino acids with the one you are looking at. So you get something like
[ARG190-CB, VAL375-N] 1.0
But you will never get something like
[VAL375-CB, VAL375-N] 1.0

I was curious if there is a way to include the atom contacts within the same residue (while the N interaction with backbones change probably won't be much, side chains, especially long ones like Lysine, can undergo significant rearrangement relative to the amide backbone).

**Comment by sroet** [comment]
Disclaimer: Currently on mobile, and have not looked at the code for more than a year. 

From a quick browse of the code, giving `n_neighbors_ignored=-1` might work. 
Setting it to `0` is definitely supported but still excludes self interactions, but `-1` might do what you want. 

It is very much untested and might error out or give unexpected/wrong results, so please check your results if it works

**Comment by sam-mahdi** [comment]
Hi sroet, 

Thank you for the assistance. Indeed setting it to -1 does work, however I appear to have noticed an issue with how this works. Just from testing, n_neighbors_ignored > 0 also appears to ignore nearby contacts. Using my current example of looking at contact map differences (for R379 query is for N, hash is for all atoms), if I use n_neighbors_ignored > 0 (i.e. 1 or the default 2), I get no contacts (this is a loop region pointing out, so it makes sense). However, if I set it to 0, I get it: 
([ARG379-N, MET378-CG], 1.0)
([ARG379-N, MET378-CB], 0.0)
([ARG379-N, MET378-N], 0.0)
([ARG379-N, MET378-CA], 0.0)
([ARG379-N, MET378-O], 0.0)
([ARG379-N, LEU380-N], 0.0)
([ARG379-N, MET378-C], 0.0)

These are not within the same chain however. These are for the i-1 and i+1 residues, so it is a bit odd they are excluded using the n_neighbors_ignored option. Setting it to -1 does indeed provide distances within the same residue

([ARG379-N, ARG379-CD], 1.0)
([ARG379-N, MET378-CG], 1.0)
([ARG379-N, MET378-CB], 0.0)
([ARG379-N, ARG379-CG], 0.0)
([ARG379-N, MET378-N], 0.0)
([ARG379-N, ARG379-CA], 0.0)
([ARG379-N, LEU380-N], 0.0)
([ARG379-N, MET378-C], 0.0)
([ARG379-N, ARG379-O], 0.0)
([ARG379-N, MET378-CA], 0.0)
([ARG379-N, MET378-O], 0.0)
([ARG379-N, ARG379-C], 0.0)
([ARG379-N, ARG379-CB], 0.0)

**Comment by sroet** [comment]
> Thank you for the assistance. Indeed setting it to -1 does work, however I appear to have noticed an issue with how this works.

Glad that worked.

> These are not within the same chain however

Just to make sure we are using the same nomenclature: A chain for us is a `mdtraj.Topology.Chain`, which normally is a single **chain** of **residues** (often one protein). Also, our flag on `n_neighbors_ignored` acts on a **residue** level, not the **atom** level (as explained [here](https://github.com/dwhswenson/contact_map/blob/master/examples/changing_defaults.ipynb) under the header "Changing the number of ignored neighbors").

With this meaning of **chain** and the flag `n_neighbors_ignored`, I assume you mean that `ARG379` is not in the same protein chain as either `MET378` or `LEU380`, so you would expect them to show up even with `n_neighbors_ignored=2`

Our logic for checking this [is here](https://github.com/dwhswenson/contact_map/blob/7fbb0b037db01831f8fd9ddd8710e9e0d23a25c5/contact_map/contact_map.py#L40-L59). 
In short for each residue:
1. it lists the `residue.index` +- `n_neighbors_ignored` (`residue.index` are unique in `mdtraj`)
2. it then makes a list of all indices of all residues in the same (protein) chain (according to the `mdtraj.topology`)
3. it only returns indices that exist **both** the list of step 1 and the list of step 2

So, if you are having issues with residues being excluded while being part of a different (protein) chain, it means your `mdtraj.Topology` is broken as it claims they **do** belong to the same chain. There is unfortunately very little we can do about that.

**Comment by sam-mahdi** [comment]
I apologize, I used the wrong terminology. This is a single protomer, so only a single chain (by chain I meant same residue).

**Comment by sroet** [comment]
> I apologize, I used the wrong terminology. This is a single protomer, so only a single chain (by chain I meant same residue).

Ah, glad that was cleared up. 

In that case I don't think there is any issue left, so I will close this. Feel free to reopen if you think that is is not the case.

**Comment by sam-mahdi** [comment]
My original question is resolved, however my follow up I'm a touch confused. When you say

`residue.index +- n_neighbors_ignored`

Is this basically ignoring +/- 2 residues of the query? I.E. If sequence is ILATI and current query is the alanine (A). Does this ignore everything else in the sequence (e.g. IL and TI)

**Comment by sroet** [comment]
> My original question is resolved, however my follow up I'm a touch confused. When you say
> 
> `residue.index +- n_neighbors_ignored`
> 
> Is this basically ignoring +/- 2 residues of the query? I.E. If sequence is ILATI and current query is the alanine (A). Does this ignore everything else in the sequence (e.g. IL and TI)
> 

Exactly. 

The original intended usecase for this package was to quickly investigate protein contacts using the default query and haystack. With those defaults you would expect these neighbor contacts to always/very often exist , but not really tell you anything interesting. 

However, if those contacts are interesting for you, you are ofcourse free to change the default (as you already do for the query), even down to -1 to include self interactions.
